enarx-keep-sgx-shim/src/main.rs

```diff
@@ -21,6 +21,59 @@ mod handler;
 use handler::Handler;
 use nolibc::x86_64::syscall::Number as SysCall;
 use sgx_types::{ssa::StateSaveArea, tcs::Tcs};
+use crt0stack::{Builder, Handle, OutOfSpace, Entry};
+
+// Hard-code the entry point for now.
+//
+// I'd like to improve this situation in the future. However, we cannot let
+// the host arbitrarily choose an entry point. Therefore, we have to pass
+// the entry point as a measured parameter to the enclave. Further, for
+// enclave portability we probably want this to be an offset from the enclave
+// starting position rather than an absolute address.
+const ENTRY: u64 = 0x2000_4000_0000;
+
+fn random() -> u64 {
+    let mut r: u64 = 0;
+
+    loop {
+        if unsafe { core::arch::x86_64::_rdrand64_step(&mut r) } == 1 {
+            return r;
+        }
+    }
+}
+
+fn crt0setup(crt0: &mut [u8]) -> Result<Handle, OutOfSpace> {
+    //let rand = unsafe { core::mem::transmute([random(), random()]) };
+
+    // Set the arguments
+    let mut builder = Builder::new(crt0);
+    builder.push("/init")?;
+
+    // Set the environement
+    let mut builder = builder.done()?;
+    builder.push("LANG=C")?;
+
+    // Set the aux vector
+    let mut builder = builder.done()?;
+    builder.push(&Entry::ExecFilename("/init"))?;
+    builder.push(&Entry::Platform("x86_64"))?;
+    builder.push(&Entry::Uid(1000))?;
+    builder.push(&Entry::EUid(1000))?;
+    builder.push(&Entry::Gid(1000))?;
+    builder.push(&Entry::EGid(1000))?;
+    builder.push(&Entry::PageSize(4096))?;
+    builder.push(&Entry::Secure(false))?;
+    builder.push(&Entry::ClockTick(100))?;
+    builder.push(&Entry::Flags(0))?;  // TODO
+    builder.push(&Entry::HwCap(0))?;  // TODO
+    builder.push(&Entry::HwCap2(0))?; // TODO
+    builder.push(&Entry::PHdr(0))?;   // TODO
+    builder.push(&Entry::PHent(0))?;  // TODO
+    builder.push(&Entry::PHnum(0))?;  // TODO
+    //builder.push(&Entry::Random(rand))?; // Currently GPFs. Not sure why...
```

Comment:
File a debt issue and put the link here if we want to merge this in commented.

Reply:
Done.